File: scripts/02_exploratory_analysis.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

# ── Setup ──────────────────────────────────────────────────────────────────────
DATA_PATH  = "data/processed/all_genes_tidy.csv"
PLOT_DIR   = "plots/exploratory"
OUTPUT_DIR = "output"
os.makedirs(PLOT_DIR, exist_ok=True)
os.makedirs(OUTPUT_DIR, exist_ok=True)

GENES = ["APOE", "CYP2C19", "HLA-B"]
GENE_COLORS = {"APOE": "#E63946", "CYP2C19": "#457B9D", "HLA-B": "#2A9D8F"}
POP_SHORT = {
    "African/African American": "AFR", "Admixed American": "AMR",
    "Ashkenazi Jewish": "ASJ", "East Asian": "EAS",
    "European (Finnish)": "FIN", "Middle Eastern": "MID",
    "European (non-Finnish)": "NFE", "Amish": "AMI",
    "South Asian": "SAS", "Remaining": "REM",
}

# ── Load data ──────────────────────────────────────────────────────────────────
df      = pd.read_csv(DATA_PATH)
overall = df[df["Population"] == "Overall"].dropna(subset=["AF"])
by_pop  = df[df["Population"] != "Overall"].dropna(subset=["AF"])

# ── 1. Summary statistics → output/ ───────────────────────────────────────────
stats = (
    overall.groupby("Gene")["AF"]
    .describe(percentiles=[0.01, 0.05, 0.25, 0.5, 0.75, 0.95, 0.99])
    .round(8)
)
stats.to_csv(f"{OUTPUT_DIR}/summary_statistics_overall.csv")
print(f"Saved → {OUTPUT_DIR}/summary_statistics_overall.csv")

pop_stats = (
    by_pop.groupby(["Gene", "Population"])["AF"]
    .describe(percentiles=[0.25, 0.5, 0.75])
    .round(8)
)
pop_stats.to_csv(f"{OUTPUT_DIR}/summary_statistics_by_population.csv")
print(f"Saved → {OUTPUT_DIR}/summary_statistics_by_population.csv")

print("\n" + "=" * 60)
print("RARE VARIANT PREVALENCE (AF < 0.01)")
print("=" * 60)
for gene in GENES:
    g     = overall[overall["Gene"] == gene]
    rare  = (g["AF"] < 0.01).sum()
    total = len(g)
    print(f"  {gene:8s}: {rare:>4d} / {total} variants are rare  ({100*rare/total:.1f}%)")

# ── 2. Plot 1: linear overview + per-gene log-scale histograms (FIXED) ─────────
fig = plt.figure(figsize=(16, 10))
fig.suptitle("Overall AF Distribution per Gene", fontsize=14, fontweight="bold")

# Top: linear scale combined
ax_lin = fig.add_subplot(2, 1, 1)
for gene in GENES:
    data = overall[overall["Gene"] == gene]["AF"]
    ax_lin.hist(data, bins=80, alpha=0.6, color=GENE_COLORS[gene], label=gene)
ax_lin.set_title("Linear Scale — all three genes")
ax_lin.set_xlabel("Allele Frequency")
ax_lin.set_ylabel("Number of Variants")
ax_lin.legend()

# Bottom: separate log-scale subplot per gene
for i, gene in enumerate(GENES):
    ax = fig.add_subplot(2, 3, 4 + i)
    data = overall[(overall["Gene"] == gene) & (overall["AF"] > 0)]["AF"]
    log_bins = np.logspace(np.log10(data.min()), np.log10(data.max()), 50)
    ax.hist(data, bins=log_bins, color=GENE_COLORS[gene], alpha=0.85, edgecolor="white")
    ax.set_xscale("log")
    ax.set_title(f"{gene} — Log Scale", fontsize=11, fontweight="bold")
    ax.set_xlabel("Allele Frequency (log scale)")
    ax.set_ylabel("Number of Variants")

plt.tight_layout()
plt.savefig(f"{PLOT_DIR}/01_overall_AF_distribution.png", dpi=150)
plt.close()
print(f"\nSaved → {PLOT_DIR}/01_overall_AF_distribution.png")

# ── 3. Plot 2: AF by population — boxplots ─────────────────────────────────────
fig, axes = plt.subplots(1, 3, figsize=(18, 6))
fig.suptitle("AF Distribution by Population and Gene (Log Scale)", fontsize=14, fontweight="bold")

for ax, gene in zip(axes, GENES):
    gene_data = by_pop[(by_pop["Gene"] == gene) & (by_pop["AF"] > 0)]
    pop_order = (
        gene_data.groupby("Population")["AF"]
        .median().sort_values(ascending=False).index.tolist()
    )
    pop_data     = [np.log10(gene_data[gene_data["Population"] == p]["AF"].values) for p in pop_order]
    short_labels = [POP_SHORT.get(p, p) for p in pop_order]

    ax.boxplot(pop_data, tick_labels=short_labels, patch_artist=True,
               boxprops=dict(facecolor=GENE_COLORS[gene], alpha=0.6),
               medianprops=dict(color="black", linewidth=2))
    ax.set_title(gene, fontsize=13, fontweight="bold")
    ax.set_xlabel("Population")
    ax.set_ylabel("log₁₀(AF)")
    ax.tick_params(axis="x", rotation=45)

plt.tight_layout()
plt.savefig(f"{PLOT_DIR}/02_AF_by_population.png", dpi=150)
plt.close()
print(f"Saved → {PLOT_DIR}/02_AF_by_population.png")

# ── 4. Plot 3: frequency class breakdown ───────────────────────────────────────
bins   = [0, 0.001, 0.01, 0.05, 0.5, 1.0]
labels = ["<0.1%", "0.1–1%", "1–5%", "5–50%", ">50%"]

fig, axes = plt.subplots(1, 3, figsize=(15, 5))
fig.suptitle("Variant Frequency Class Breakdown per Gene", fontsize=14, fontweight="bold")

for ax, gene in zip(axes, GENES):
    data   = overall[overall["Gene"] == gene]["AF"].dropna()
    data   = data[data > 0]
    cats   = pd.cut(data, bins=bins, labels=labels, include_lowest=True)
    counts = cats.value_counts().reindex(labels).fillna(0).astype(int)

    bars = ax.bar(labels, counts.values, color=GENE_COLORS[gene], alpha=0.8, edgecolor="white")
    ax.set_yscale("log")
    ax.set_ylim(bottom=0.5)
    ax.set_title(gene, fontsize=13, fontweight="bold")
    ax.set_xlabel("Frequency Class")
    ax.set_ylabel("Number of Variants (log scale)")
    ax.tick_params(axis="x", rotation=30)

    for bar, count in zip(bars, counts.values):
        if count > 0:
            ax.text(bar.get_x() + bar.get_width() / 2, bar.get_height() * 1.2,
                    str(count), ha="center", va="bottom", fontsize=9, fontweight="bold")

plt.tight_layout()
plt.savefig(f"{PLOT_DIR}/03_frequency_class_breakdown.png", dpi=150)
plt.close()
print(f"Saved → {PLOT_DIR}/03_frequency_class_breakdown.png")

print("\nDone. All outputs saved.")

File: R_version/scripts/02_exploratory_analysis.R
library(tidyverse)
library(gridExtra)
library(grid)

# ── Setup ──────────────────────────────────────────────────────────────────────
DATA_PATH  <- "data/processed/all_genes_tidy.csv"
PLOT_DIR   <- "data/plots/exploratory"
OUTPUT_DIR <- "data/output"
dir.create(PLOT_DIR, recursive = TRUE, showWarnings = FALSE)
dir.create(OUTPUT_DIR, recursive = TRUE, showWarnings = FALSE)

GENE_COLORS <- c("APOE" = "#E63946", "CYP2C19" = "#457B9D", "HLA-B" = "#2A9D8F")
POP_SHORT <- c(
  "African/African American" = "AFR", "Admixed American" = "AMR",
  "Ashkenazi Jewish" = "ASJ",         "East Asian" = "EAS",
  "European (Finnish)" = "FIN",       "Middle Eastern" = "MID",
  "European (non-Finnish)" = "NFE",   "Amish" = "AMI",
  "South Asian" = "SAS",              "Remaining" = "REM"
)

df      <- read_csv(DATA_PATH)
overall <- df %>% filter(Population == "Overall") %>% drop_na(AF)
by_pop  <- df %>% filter(Population != "Overall") %>% drop_na(AF)

# ── 1. Summary Statistics ──────────────────────────────────────────────────────
summary_overall <- overall %>%
  group_by(Gene) %>%
  summarise(
    count = n(), mean = mean(AF), std = sd(AF), min = min(AF),
    `1%` = quantile(AF, 0.01), `5%` = quantile(AF, 0.05), `25%` = quantile(AF, 0.25),
    `50%` = quantile(AF, 0.50), `75%` = quantile(AF, 0.75), `95%` = quantile(AF, 0.95),
    `99%` = quantile(AF, 0.99), max = max(AF)
  )
write_csv(summary_overall, file.path(OUTPUT_DIR, "summary_statistics_overall.csv"))

summary_pop <- by_pop %>%
  group_by(Gene, Population) %>%
  summarise(
    count = n(), mean = mean(AF), std = sd(AF), min = min(AF),
    `25%` = quantile(AF, 0.25), `50%` = quantile(AF, 0.50), `75%` = quantile(AF, 0.75), max = max(AF)
  )
write_csv(summary_pop, file.path(OUTPUT_DIR, "summary_statistics_by_population.csv"))

message("\n============================================================")
message("RARE VARIANT PREVALENCE (AF < 0.01)")
message("============================================================")
overall %>%
  group_by(Gene) %>%
  summarise(rare = sum(AF < 0.01), total = n()) %>%
  purrr::pwalk(function(Gene, rare, total) {
    cat(sprintf("  %-8s: %4d / %d variants are rare  (%.1f%%)\n", Gene, rare, total, 100 * rare / total))
  })

# ── 2. Plot 1: Linear overview + Log Histograms ─────────────────────────────────
library(gridExtra)
p_lin <- ggplot(overall, aes(x = AF, fill = Gene)) +
  geom_histogram(bins = 80, alpha = 0.6, position = "identity") +
  scale_fill_manual(values = GENE_COLORS) +
  labs(title = "Linear Scale — all three genes", x = "Allele Frequency", y = "Number of Variants") +
  theme_minimal()

log_plots <- overall %>%
  filter(AF > 0) %>%
  group_split(Gene) %>%
  map(function(sub) {
    gene_name <- unique(sub$Gene)
    ggplot(sub, aes(x = AF)) +
      geom_histogram(bins = 50, fill = GENE_COLORS[gene_name], color = "white", alpha = 0.85) +
      scale_x_log10() +
      labs(title = paste(gene_name, "— Log Scale"), x = "Allele Frequency (log scale)", y = "Number of Variants") +
      theme_minimal() +
      theme(plot.title = element_text(face = "bold"))
  })

png(file.path(PLOT_DIR, "01_overall_AF_distribution.png"), width = 16, height = 10, units = "in", res = 150)
grid.arrange(p_lin, do.call(arrangeGrob, c(log_plots, ncol = 3)), nrow = 2, 
             top = textGrob("Overall AF Distribution per Gene", gp = gpar(fontsize = 14, fontface = "bold")))
dev.off()

# ── 3. Plot 2: Boxplots by Population ───────────────────────────────────────────
box_plots <- by_pop %>%
  filter(AF > 0) %>%
  group_split(Gene) %>%
  map(function(sub) {
    gene_name <- unique(sub$Gene)
    
    # Calculate median sorting logic matching your Python workflow
    pop_order <- sub %>%
      group_by(Population) %>%
      summarise(med = median(AF)) %>%
      arrange(desc(med)) %>%
      pull(Population)
    
    sub_ordered <- sub %>% 
      mutate(Population = factor(Population, levels = pop_order))
    
    ggplot(sub_ordered, aes(x = Population, y = log10(AF))) +
      geom_boxplot(fill = GENE_COLORS[gene_name], alpha = 0.6, outlier.size = 0.5) +
      scale_x_discrete(labels = POP_SHORT) +
      labs(title = gene_name, x = "Population", y = "log10(AF)") +
      theme_minimal() +
      theme(
        axis.text.x = element_text(angle = 45, hjust = 1), # <-- Standard clean R text rotation
        plot.title = element_text(face = "bold")
      )
  })

png(file.path(PLOT_DIR, "02_AF_by_population.png"), width = 18, height = 6, units = "in", res = 150)
grid.arrange(do.call(arrangeGrob, c(box_plots, ncol = 3)), 
             top = textGrob("AF Distribution by Population and Gene (Log Scale)", gp = gpar(fontsize = 14, fontface = "bold")))
dev.off()

# ── 4. Plot 3: Frequency Class Breakdown ────────────────────────────────────────
bins_classes <- c(0, 0.001, 0.01, 0.05, 0.5, 1.0)
labels_classes <- c("<0.1%", "0.1–1%", "1–5%", "5–50%", ">50%")

class_plots <- overall %>%
  filter(AF > 0) %>%
  mutate(Class = cut(AF, breaks = bins_classes, labels = labels_classes, include.lowest = TRUE)) %>%
  group_split(Gene) %>%
  map(function(sub) {
    gene_name <- unique(sub$Gene)
    counts <- sub %>% group_by(Class) %>% tally()
    
    ggplot(counts, aes(x = Class, y = n)) +
      geom_bar(stat = "identity", fill = GENE_COLORS[gene_name], alpha = 0.8, color = "white") +
      geom_text(aes(label = n), vjust = -0.5, fontface = "bold", size = 3) +
      scale_y_log10(limits = c(0.5, max(counts$n) * 5)) +
      labs(title = gene_name, x = "Frequency Class", y = "Number of Variants (log scale)") +
      theme_minimal() +
      theme(axis.text.x = element_text(angle = 30, hjust = 1), plot.title = element_text(face = "bold"))
  })

png(file.path(PLOT_DIR, "03_frequency_class_breakdown.png"), width = 15, height = 5, units = "in", res = 150)
grid.arrange(do.call(arrangeGrob, c(class_plots, ncol = 3)), 
             top = textGrob("Variant Frequency Class Breakdown per Gene", gp = gpar(fontsize = 14, fontface = "bold")))
dev.off()